### PR TITLE
forum: Add ReadState. Fixes #81

### LINKED ISSUE #81: forum: add a way to check if a thread has new unread messages
> ###### Originally posted 2017-02-07 11:35 by [serialk](https://bitbucket.org/serialk/) on Bitbucket

### PR DESCRIPTION
**Pull request by cafehaine**
TODO:
- mark thread as read only if viewing the last page

### CODE REVIEW

File: prologin/forum/views.py [hunk 1]
```diff
@@ -214,6 +224,21 @@ def get(self, request, *args, **kwargs):
...
                             pk=thread.pk)
+
+        # if user isn't anonymous, mark thread as read
+        if not self.request.user.is_anonymous:
+            try:
+                lastviewed = forum.models.LastViewed.objects.get(user=self.request.user,
+                                                                 thread=thread)
+                lastviewed.post = thread.last_post
+                lastviewed.save()
+            except forum.models.LastViewed.DoesNotExist:
```
Comment (zopieux) [inline prologin/forum/views.py]: Seems like the exact use case of [`update_or_create`](https://docs.djangoproject.com/en/2.0/ref/models/querysets/#update-or-create).

File: prologin/forum/views.py [hunk 2]
```diff
@@ -119,6 +119,16 @@ def get_context_data(self, **kwargs):
         context['sticky_threads'] = [thread for thread in threads if thread.is_sticky]
         context['normal_threads'] = threads[len(context['sticky_threads']):]
         context['announces'] = self.get_forum.threads.filter(type=forum.models.Thread.Type.announce.value)
+        # if user is logged in
+        if not self.request.user.is_anonymous:
```
Comment (seirl) [inline prologin/forum/views.py]: if self.request.user.is_authenticated

File: prologin/forum/models.py [hunk 3]
```diff
@@ -273,6 +273,18 @@ def get_permalink(self):
         return reverse('forum:post', kwargs={'thread_slug': self.thread.slug, 'pk': self.pk})
 
 
+class LastViewed(models.Model):
+    thread = models.ForeignKey(Thread, related_name='user_last_viewed',
+                               on_delete=models.CASCADE)
+
+    post = models.ForeignKey(Post, related_name='user_last_viewed',
```
Comment (seirl) [inline prologin/forum/models.py]: Move this at the end of the class to be clear on the fact that the manytomany is between threads and users.

File: prologin/forum/models.py [hunk 4]
```diff
@@ -273,6 +273,18 @@ def get_permalink(self):
         return reverse('forum:post', kwargs={'thread_slug': self.thread.slug, 'pk': self.pk})
 
 
+class LastViewed(models.Model):
```
Comment (seirl) [inline prologin/forum/models.py]: Rename the class in ReadState, it conveys the intention better.

File: prologin/forum/models.py [hunk 5]
```diff
@@ -273,6 +273,18 @@ def get_permalink(self):
...
+    thread = models.ForeignKey(Thread, related_name='user_last_viewed',
```
Comment (seirl) [inline prologin/forum/models.py]: related_name='read_states'

File: prologin/forum/models.py (hunk 3 above)
Comment (seirl) [reply prologin/forum/models.py]: related_name='read_states'

File: prologin/forum/models.py [hunk 6]
```diff
@@ -273,6 +273,18 @@ def get_permalink(self):
...
+                               on_delete=models.CASCADE)
+
+    post = models.ForeignKey(Post, related_name='user_last_viewed',
+                             on_delete=models.CASCADE)
+
+    user = models.ForeignKey(settings.AUTH_USER_MODEL,
+                             related_name='forum_last_viewed_post',
```
Comment (seirl) [inline prologin/forum/models.py]: related_name='read_states'

File: prologin/forum/views.py [hunk 7]
```diff
@@ -214,6 +224,21 @@ def get(self, request, *args, **kwargs):
                             forum_pk=thread.forum.pk,
                             slug=thread.slug,
                             pk=thread.pk)
+
+        # if user isn't anonymous, mark thread as read
+        if not self.request.user.is_anonymous:
+            try:
+                lastviewed = forum.models.LastViewed.objects.get(user=self.request.user,
+                                                                 thread=thread)
+                lastviewed.post = thread.last_post
```
Comment (seirl) [inline prologin/forum/views.py]: Move those two lines after the except: block to factor them out.

File: prologin/forum/views.py [hunk 8]
```diff
@@ -214,6 +224,21 @@ def get(self, request, *args, **kwargs):
...
+                lastviewed.save()
+            except forum.models.LastViewed.DoesNotExist:
+                lastviewed = forum.models.LastViewed()
```
Comment (seirl) [inline prologin/forum/views.py]: LastViewed(user=self.request.user, thread=thread)

File: prologin/forum/views.py [hunk 9]
```diff
@@ -119,6 +119,16 @@ def get_context_data(self, **kwargs):
         context['sticky_threads'] = [thread for thread in threads if thread.is_sticky]
         context['normal_threads'] = threads[len(context['sticky_threads']):]
         context['announces'] = self.get_forum.threads.filter(type=forum.models.Thread.Type.announce.value)
+
+        # if user is logged in fetch read threads
+        if self.request.user.is_authenticated:
```
Comment (seirl) [inline prologin/forum/views.py]: We don't want to have that here, but rather do it in the model manager. Something like (untested):

```
def seen_by(self, user=None):
    return (self.get_queryset()
            .select_related('read_states')
            .filter(read_states__user=user)
            .annotate(seen=Case(
                When(post=F('read_states__post'), then=Value(1)),
                default=Value(0)
                output_field=BooleanField()
            )))

```

File: prologin/forum/templates/forum/stub_forum_thread.html [hunk 10]
```diff
@@ -10,7 +10,7 @@
         {% if thread.is_closed %}<i class="fa fa-lg fa-lock"></i>{% endif %}
       </div>
       <a href="{{ thread.get_absolute_url }}">
-        <h3>{{ thread.title }}</h3>
+      <h3>{{ thread.title }}{% if not thread in read_threads %}*{% endif %}</h3>
```
Comment (seirl) [inline prologin/forum/templates/forum/stub_forum_thread.html]: Use \<strong> for unread threads

**Review by seirl — CHANGES_REQUESTED** [review]

**Comment by seirl** [comment]
We should only mark the thread as read if the user looks at the last page of the thread, not any page.

**Review by zopieux — CHANGES_REQUESTED** [review]

File: prologin/forum/views.py (hunk 1 above)
Comment (cafehaine) [reply prologin/forum/views.py]: didn't know this existed, thanks !

File: prologin/forum/models.py [hunk 11]
```diff
@@ -273,6 +273,18 @@ def get_permalink(self):
...
+                             on_delete=models.CASCADE)
+
```
Comment (seirl) [inline prologin/forum/models.py]: You also need a unique_together between user and thread.

File: prologin/forum/templates/forum/stub_forum_thread.html [hunk 12]
```diff
@@ -10,7 +10,8 @@
         {% if thread.is_closed %}<i class="fa fa-lg fa-lock"></i>{% endif %}
       </div>
       <a href="{{ thread.get_absolute_url }}">
-        <h3>{{ thread.title }}</h3>
+        {% if thread.seen %}<h3>{{ thread.title }}</h3>
```
Comment (seirl) [inline prologin/forum/templates/forum/stub_forum_thread.html]: Maybe just <h3 {% if thread.seen %}class="thread-unread"{% endif %}> with the corresponding font-weight: bold; entry

File: prologin/forum/views.py [hunk 13]
```diff
@@ -119,6 +119,7 @@ def get_context_data(self, **kwargs):
         context['sticky_threads'] = [thread for thread in threads if thread.is_sticky]
         context['normal_threads'] = threads[len(context['sticky_threads']):]
         context['announces'] = self.get_forum.threads.filter(type=forum.models.Thread.Type.announce.value)
+
```
Comment (seirl) [inline prologin/forum/views.py]: remove

**Review by seirl — CHANGES_REQUESTED** [review]

**Pull request merged**